apps/prairielearn/src/sprocs/sync_assessments.sql

```diff
@@ -525,12 +525,26 @@ BEGIN
     FROM (
         SELECT
             tid,
-            row_number() OVER (ORDER BY (
-                SELECT string_agg(convert_to(coalesce(r[2],
-                    length(length(r[1])::text) || length(r[1])::text || r[1]),
-                    'SQL_ASCII'),'\x00')
-                FROM regexp_matches(number, '0*([0-9]+)|([^0-9]+)', 'g') r
-            ) ASC) AS order_by
+            row_number() OVER (
+                ORDER BY (
+                    SELECT
+                        string_agg(
+                            convert_to(
+                                coalesce(
+                                    r[2],
+                                    length(length(r[1])::text) || length(r[1])::text || r[1]
+                                ),
+                                'SQL_ASCII'
+                            ),
+                            '\x00'
+                        )
+                    FROM
+                        regexp_matches(number, '0*([0-9]+)|([^0-9]+)', 'g') r
+                ) ASC,
+                -- In case two assessments have the same number, fall back to
+                -- ordering by the ID to ensure a stable sort.
+                id ASC
```

Comment:
In case it wasn't obvious (I had to double check), Postgres doesn't at all guarantee sort order when an `ORDER BY` clause does not uniquely order all elements:

> If sorting is not chosen, the rows will be returned in an unspecified order. The actual order in that case will depend on the scan and join plan types and the order on disk, but it must not be relied on. ([source](https://www.postgresql.org/docs/current/queries-order.html))